### PR TITLE
 Hooks: always use fastify core objects

### PR DESCRIPTION
**Pull request by delvedor**
**Why this PR?**
Have `onRequest` and `onResponse` with different signature and objects leads to confusion and slightly complex internals, this pr fixes that.

**What changes?**
`onRequest` will have the same objects as `preHandler`, the difference is that the `body` has not been parsed yet.
`onResponse` will also get the full request object (which could be useful) and the reply object.
Middlewares will continue to use the Node core objects.

```js
  fastify.addHook('onRequest', function (request, reply, next) {
    // request and reply are the core Fastify objects
    next()
  })

  fastify.addHook('onResponse', function (request, reply, next) {
	// request and reply are the core Fastify objects
    next()
  })

  fastify.use(function (req, res, next) {
    // req and res are the Node core objects
    next()
  })
```

This is a *breaking change*, so if accepted it will land in the next semver major.

Feedbacks? :)

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message and code follows [Code of conduct](https://github.com/fastify/fastify/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

File: docs/Ecosystem.md [hunk 1]
```diff
@@ -75,6 +75,7 @@ Plugins maintained by the fastify team are listed under [Core](#core) while plug
 - [`fastify-soap-client`](https://github.com/fastify/fastify-soap-client) a SOAP client plugin for Fastify.
 - [`fastify-sse`](https://github.com/lolo32/fastify-sse) to provide Server-Sent Events with `reply.sse( … )` to Fastify.
 - [`fastify-tls-keygen`](https://gitlab.com/sebdeckers/fastify-tls-keygen) Automatically generate a browser-compatible, trusted, self-signed, localhost-only, TLS certificate.
+- [`fastify-vhost`](https://github.com/patrickpissurno/fastify-vhost) Proxy subdomain http requests to another server (useful if you want to point multiple subdomains to the same IP address, while running different servers on the same machine).
```
Comment (allevo) [inline docs/Ecosystem.md]: Could this change be made on a separated PR?

Comment (delvedor) [reply docs/Ecosystem.md]: Done ;)

**Comment by jsumners** [comment]
I am very much in support of this change. It has frustrated me for quite some time.

File: fastify.js [hunk 2]
```diff
@@ -226,45 +222,23 @@ function build (options) {
...
     req.log.info({ req }, 'incoming request')
 
-    if (hasLogger === true || context.onResponse !== null) {
-      res.on('finish', onResFinished)
-      res.on('error', onResFinished)
-    }
+    var request = new context.Request(params, req, urlUtil.parse(req.url, true).query, req.headers, req.log)
+    var reply = new context.Reply(res, context, request, res.log)
+
+    reply._setup(hasLogger)
```
Comment (mcollina) [inline fastify.js]: I’m not fond of this. Can you put it behind a symbol, or make it a top level function (whichever is faster).

File: lib/reply.js [hunk 3]
```diff
@@ -326,8 +336,37 @@ function handleError (reply, error, cb) {
...
+
+  var onResFinished = err => {
+    this.res.removeListener('finish', onResFinished)
+    this.res.removeListener('error', onResFinished)
+
+    var ctx = this.context
+
+    if (ctx && ctx.onResponse !== null) {
+      // deferring this with setImmediate will
+      // slow us by 10%
```
Comment (mcollina) [inline lib/reply.js]: Can you remove this comment?

**Comment by nwoltman** [comment]
Definitely in favour of this change 👍

**Comment by allevo** [comment]
Can you run the benchmarks?

**Comment by delvedor** [comment]
@allevo I'll run the benchmarks asap, but I don't expect changes. Probably we will go even faster since we are no longer creating the state object and we are altering less the Node core objects.

In https://github.com/fastify/fastify/pull/1066/commits/9191003158c46efe28003adcfc6ff398cdaa961c, I did some housekeeping, I've merged `lib/hookRunner.js` and `lib/hooks.js` in the same file since keep them separated were useless.
I've also updated some internal function and possibly improved the performances because we are accessing even fewer objects.

File: fastify.js (hunk 2 above)
Comment (delvedor) [reply fastify.js]: Done!

**Comment by mcollina** [comment]
Can you start assembling a "notable changes"/breaking changes list for fastify 2 somewhere?

**Pull request merged**